cleverhans/future/tf2/utils_tf.py

```diff
@@ -108,3 +108,48 @@ def random_lp_vector(shape, ord, eps, dtype=tf.float32, seed=None):
     r = eps * tf.reshape(w * x / norm, shape)
 
   return r
+
+def get_or_guess_labels(model_fn, x, y=None, targeted=False):
```

Comment:
Sounds inefficient to recompute predictions when y is given: if my understanding is correct v1 was just copying in that case. Also I am not convinced this is useful as a standalone function, the code seemed much more natural to me in v1 (l.249 to 298 in (https://github.com/cleverhans-lab/cleverhans/blob/master/cleverhans/attacks/carlini_wagner_l2.py).

Reply:
The original code uses the tf1 version as well (https://github.com/cleverhans-lab/cleverhans/blob/master/cleverhans_v3.1.0/cleverhans/attacks/carlini_wagner_l2.py#L70).

My implementation is a port of the original one. I simply added that you can also supply the labels in categorical encoding. The torch version has a similar function in their utility file (https://github.com/cleverhans-lab/cleverhans/blob/07cfa55c0b34da7b9bd11b0b8955da5422b4eca0/cleverhans/torch/utils.py#L42-L66).